File: java/023. Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length>2) {
            ListNode[] l1=new ListNode[lists.length/2];
            ListNode[] l2=new ListNode[lists.length-lists.length/2];
            for (int i = 0; i < lists.length; i++) {
                if (i<lists.length/2) {
                    l1[i]=lists[i];
                }else {
                    l2[i-lists.length/2]=lists[i];
                }
            }
            return mergeTwoLists(mergeKLists(l1),mergeKLists(l2));
        }else if (lists.length==2){
            return mergeTwoLists(lists[0],lists[1]);
        }else if (lists.length==1){
            return lists[0];
        }else {
            return null;
        }
    }
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {  
        if (l1==null) {  
            return l2;  
        }  
        if (l2==null) {  
            return l1;  
        }  
        ListNode l=null;  
        if (l1.val>l2.val) {  
            l=new ListNode(l2.val);  
            l.next=mergeTwoLists(l1,l2.next);  
        }else {  
            l=new ListNode(l1.val);  
            l.next=mergeTwoLists(l1.next,l2);  
        }  
        return l;  
    }  
}

File: ruby/023. Merge k Sorted Lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
    l=lists.length
    if l<=2
        return merge_two_lists(lists[0],l==2 ? lists[1] : nil)
    end
    return merge_two_lists(merge_k_lists(lists[0,l/2]),merge_k_lists(lists[l/2,l-l/2]))
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def merge_two_lists(l1, l2)
    if not l1
        return l2
    end
    if not l2
        return l1
    end
    if l1.val > l2.val
        return merge_two_lists(l2, l1)
    end
    l1.next=merge_two_lists(l1.next,l2)
    return l1
end